### ISSUE TITLE
GEOSgcm GNU runs do not pass regression

### ISSUE THREAD

**Issue by mathomp4** [GEOS-ESM/GEOSgcm#380]
So, now that we've fixed the GNU run issue by updating to Chem 1.8.1 (#378), my tests are now showing that *ALL* GNU runs do not pass regression. I'm not sure if it's start-stop or layout, but I have to assume layout since start-stop is usually much more fundamental.

But I'm baffled as to why Intel would. My guess is this is some pointer thing again in either Chem or GOCART. I'll probably be pinging @tclune, @atrayano and @bena-nasa to help with this. 

First, I'll try and confirm that it is layout.

**Comment by tclune** [comment]
May I just say:  "bleh".

**Comment by mathomp4** [comment]
Actually, it might be start-stop. I tried 4x24 and 3x12 and all was good. Time to make up a start-stop test.

**Comment by mathomp4** [comment]
Yeah. It's start-stop. Huh.

**Comment by mathomp4** [comment]
Okay. So with the help of @aoloso, I ran one instance at a time. And I got:

| Instance | Restart Failures |
|:--------:|:----------------:|
| DU       | du_int, ss_int |
| SS       | fvcore_int, moist_int, moist_imp, aiau_imp, turb_imp, gocart_imp, du_int, ss_int, hemco_imp, saltwater_imp |
| CA.oc    | du_int, ss_int |
| CA.bc    | du_int, ss_int |
| CA.br    | du_int, ss_int |
| SU       | du_int, ss_int, su_int |
| NI       | du_int, ss_int, ni_int |

**Comment by mathomp4** [comment]
I think I might need to ping @pcolarco to start. 

Pete, for context, the runs above were made as so:

1. Run 2 steps, dump checkpoints
2. Run 1 step, stop, run 1 step, dump checkpoints

For each run, I edited these loops in Run1:

```fortran
!   Run the children
!   -----------------
    do i = 1, size(gcs)
      call ESMF_GridCompRun (gcs(i), importState=gim(i), exportState=gex(i), phase=1, clock=clock, __RC__)
    end do
```

and Run2:

```fortran
!   Run the children
!   -----------------
    do i = 1, size(gcs)
      call ESMF_GridCompGet (gcs(i), NAME=child_name, __RC__ )
      if ((index(child_name, 'data')) == 0) then ! only execute Run2 method if a computational instance
         call ESMF_GridCompRun (gcs(i), importState=gim(i), exportState=gex(i), phase=2, clock=clock, __RC__)
      end if
    end do
```
where I changed to loop to be:
```fortran
    do i = 1, 1
...
```
and run, then did 2, 3, 4...

This was pretty much the only way @aoloso could figure out how to run just part of G2G. (Maybe there's a better way?)

What the table shows are which checkpoints are different when running *just* that instance. So for most instances, it's always DU and SS. But when only SS is run...hoo boy.

**Comment by pcolarco** [comment]
Comments/questions.
1) I've not yet tried running only a subset of the components. Some of them talk to each other (NI for example depends on what SU, SS, and DU are doing). It sounds like you are telling me you cannot run a subset without hacking the fortran, is that right?
2) If only running DU why are you looking at/getting an ss_internal_rst at all anyway? I don't understand what you are reporting your table.
3) If the aerosols are radiatively active (are they?) then turning on/off individual components will change the state of everything else in the model, including other species checkpoint values due to change in dynamics (at least). Is that what's going on?

**Comment by mathomp4** [comment]
> Comments/questions.
> 
> 1. I've not yet tried running only a subset of the components. Some of them talk to each other (NI for example depends on what SU, SS, and DU are doing). It sounds like you are telling me you cannot run a subset without hacking the fortran, is that right?

Well, I just wasn't sure *how* to run just a subset. I tried fiddling with `GOCART2G_GridComp.rc` but almost everything I tried sort of crashed the model.

> 2. If only running DU why are you looking at/getting an ss_internal_rst at all anyway? I don't understand what you are reporting your table.

What my test was was a variant of `gcm_regress.j`. The normal run is:

1. Run model for 1 day at NX x NY → Gives us Set #1 of Restarts.
2. Run model for 18 hours at NY x NX, stop.
3. Run model for 6 hours at 1x6 → Gives us Set #2 of Restarts.

The script then compares all of Set #1 to Set #2. It's a sort of "super zero-diff" test. 

Now my tests with GNU show that changing the layout (4x24 vs 6x36 or whatever) was fixed by your bugfix that fixed it for Intel as well. 

But I found that GNU *does not* handle start-stop regression. When you run with GNU, if you run for 1 day straight, you get different answers when you run for 18 hours and then 6 hours.

Further testing showed that you can see this by comparing 2 steps vs 1 step + 1 step. 

My tests above show what you get when you compare *ALL* the restarts generated by a 2-step run versus a 1-step+1-step. 

I was further then trying to figure out what component in G2G might cause this. I suspect G2G because it is the main difference from GOCART 1 days when GNU regressed just fine.

> 3. If the aerosols are radiatively active (are they?) then turning on/off individual components will change the state of everything else in the model, including other species checkpoint values due to change in dynamics (at least). Is that what's going on?

Well, I'm not comparing the model against any other configuration. I am just trying to see what happens if you run a 2 steps versus 1 step + 1 step *changing nothing*. With Intel, these two runs give the same answers, but we do not get that with GNU.

**Comment by pcolarco** [comment]
I take back part of my question. I think you can turn off a component in GOCART2G_GridComp.rc by removing its name like this, from:
ACTIVE_INSTANCES_SU:  SU
to
ACTIVE_INSTANCES_SU: 

I did this working on the previous regression test to isolate SU as the culprit. Commenting it out (#SU) does not work, you have remove from the line entirely.

Now based on what you say I wonder if there's some sort of missing "import_rst" for the various components at play here. Alternatively, could be the alarm for the SU oxidants isn't behaving as expected.

Those are my pre-dinner thoughts.

**Comment by mathomp4** [comment]
> I take back part of my question. I think you can turn off a component in GOCART2G_GridComp.rc by removing its name like this, from: ACTIVE_INSTANCES_SU: SU to ACTIVE_INSTANCES_SU:
> 
> I did this working on the previous regression test to isolate SU as the culprit. Commenting it out (#SU) does not work, you have remove from the line entirely.

Ahhh. Yeah, I think I was just commenting out since, well, there are other comments on the line. I'll try that.

> Now based on what you say I wonder if there's some sort of missing "import_rst" for the various components at play here. Alternatively, could be the alarm for the SU oxidants isn't behaving as expected.

Glad you have an idea! I'll start doing some tests removing active instances.

**Comment by mathomp4** [comment]
Well, some more tests with turning off instances in `RC/GOCART2G_GridComp.rc`:

| Instance(s) Turned Off |                                              Restart Failures                                              |
|:----------------------:|:----------------------------------------------------------------------------------------------------------:|
|           DU           |                                         All but gwd_imp, irrad_int                                         |
|         DU, CA         |                                        CRASH (line 1817 in G2G GC)                                         |
|           SS           |                                              CRASH (NI need?)                                              |
|           SU           |                                              CRASH (NI need?)                                              |
|           NI           |                                         All but gwd_imp, irrad_int                                         |
|         NI, CA         | fvcore_int, moist_int, aiau_imp, moist_imp, turb_imp, gocart_imp, du_int, su_int, hemco_imp, saltwater_imp |
|         NI, SS         |                                               du_int, su_int                                               |
|       NI, SS, DU       |                                        CRASH (line 1700 in G2G GC)                                         |
|     NI, SS, DU, SU     |                                        CRASH (line 1700 in G2G GC)                                         |
|       NI, SS, SU       |                                                   du_int                                                   |
|         NI, SU         |                                         All but gwd_imp, irrad_int                                         |
|       NI, SU, DU       |                                        CRASH (line 1700 in G2G GC)                                         |
|                        |                                                                                                            |

Not much logic here I can see.

Also, I'm beginning to think we might need some protections in GOCART saying things like "You can't run NI without SS" or something...

**Comment by pcolarco** [comment]
@mathomp4 Is that line in the aerosol activation code? (qaux = qa_)
Try making sure your AGCM.rc has this line as I paste it below:
USE_AEROSOL_NN: 0
Maybe you're crashing because the single moment cloud scheme is expecting all the aerosol species but you're not providing them.

**Comment by mathomp4** [comment]
@pcolarco Ah. Okay. Thanks. That did it. I'll try and re-run the line 1700 crashes with that option.

**Comment by mathomp4** [comment]
Well, some progress. I've updated the table and I seem to have found a way to get GNU to regress: Just run CA with aerosol activation off! 😄 

I suppose one question that maybe @pcolarco can answer: How is CA different from SS, DU, SU, and NI+SS+DU?  Is there a routine that everything *but* CA uses? I tried doing some side-by-side comparisons, but CA doesn't seem that unique. I mean, it has the weird 'three instances in one' thing, but...not much else.

**Comment by pcolarco** [comment]
@mathomp4 Disclaimer: I was not involved in the aerosol activation bit and only learned it was a thing by stumbling over it myself. My suspicion is that--since the Brown Carbon bit in CA is newish--the brown carbon is not connected properly with the aerosol activation. I will try and look at it this morning.

**Comment by mathomp4** [comment]
Okay. So after some work, I can now say I can get GNU to regress...as long as I only run for 2 steps and set `USE_AEROSOL_NN: 0`. I can do this by setting `mintracer` in `ConvPar_GF_GEOS5.F90` to zero instead of `tiny(1.)`.

Unfortunately, this value of `mintracer` has been that way since we started on Git and I couldn't figure out in CVS land where it came from. I'm pretty sure the change came from @wmputman on one of his branches, but he probably got it from Saulo at some point.

That said, I'm not exactly sure why `mintracer` isn't zero. I mean...no aerosols or whatever is valid. Negative tracers is a bad thing, but zero isn't I don't think.

**Comment by mathomp4** [comment]
Update. I was in a telecon with @tclune @weiyuan-jiang @junwang-noaa and said "Let me try GCC 11.2 and Intel MPI" because I've been experimenting with that recently. And when I ran it...it regressed!

Meanwhile, last night GCC 11.2 + Open MPI did not (as usual). Buuuuuut, there was one more difference. In my IMPI test, I updated to [my Baselibs 7.5.0/MAPL 2.23 branch of GEOSgcm](https://github.com/GEOS-ESM/GEOSgcm/pull/428) first.

So, I now have new things to try. I will report back soon.

**Comment by mathomp4** [comment]
One more update to this. It also regresses just fine at NAS with MPT (hadn't noticed this, been blind).

So, Intel MPI: Pass, MPT: Pass, Open MPI: Fail. 

Very weird that MPT works. I mean, MPT is always the one we have issues with!

This is some confusing stuff. I mean, the only MPI directly in GOCART is Bcasts of reals which can't be responsible. And the fact that it's start-stop which in a way is even *less* involved with MPI. I could imagine MPI stacks causing issues with MPI layout regression. But start-stop with no layout changes??

Well, might be time to become a beta-tester for @jsquyres, et al, and try out Open MPI 5...

**Comment by jsquyres** [comment]
Since you tagged me, I'm curious about this: did you find and resolve the issue?  Was it something to do with Open MPI?

**Comment by mathomp4** [comment]
@jsquyres In the end, no. It *seemed* like Open MPI was the cause since Intel MPI and MPT worked just fine. But I finally realized that we run with Open MPI, we run with one setting in our IO layer differently. 

Long ago, Open MPI writing our checkpoint files in our normal code path was like 1000x slower or something, but it was fine in an alternate path. This was always weird to me since, in the end, the two paths are nigh-identical but, hey, it worked.

Once I figured that out, we looked at the alternate code path and found a bug in 4d writing. Fixed that and it seems good. Now any issues I have seem to be GNU related and not Open MPI related.

Of course, that said, you will probably be hearing more from me. My first attempt with Open MPI 5 already led to a couple questions:

* https://github.com/open-mpi/ompi/issues/11299
* https://www.mail-archive.com/users@lists.open-mpi.org/msg35098.html
